examples/all-clusters-app/all-clusters-common/src/DeviceEnergyManagementDelegateImpl.cpp

```diff
@@ -0,0 +1,250 @@
+/*
+ *
+ *    Copyright (c) 2023 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "DeviceEnergyManagementDelegateImpl.h"
+// #include <chrono>
+// #include <string>
+// #include <thread>
+
+using namespace chip::app::Clusters::DeviceEnergyManagement;
+using namespace chip::app::Clusters::DeviceEnergyManagement::Attributes;
+
+using chip::Optional;
+using namespace chip::app;
+using CostsList = DataModel::List<const Structs::CostStruct::Type>;
+
+Status DeviceEnergyManagementDelegate::PowerAdjustRequest(const int64_t power, const uint32_t duration)
+{
+    Status status = Status::UnsupportedCommand; // Status::Success;
+
+    // TODO: implement
+    mEsaState = ESAStateEnum::kPowerAdjustActive;
+
+    // TODO:  Generate a PowerAdjustStart Event, then begins to adjust its power
+    // When done, raise PowerAdjustEnd & ESAState set to kOnline.
+
+    MatterReportingAttributeChangeCallback(mEndpointId, DeviceEnergyManagement::Id, ESAState::Id);
+
+    return status;
+}
+
+Status DeviceEnergyManagementDelegate::CancelPowerAdjustRequest()
+{
+    Status status = Status::UnsupportedCommand; // Status::Success;
+
+    // TODO: implement
+    /* TODO:  If the command is accepted, the ESA SHALL generate an PowerAdjustEnd Event.    */
+    mEsaState = ESAStateEnum::kOnline;
+    MatterReportingAttributeChangeCallback(mEndpointId, DeviceEnergyManagement::Id, ESAState::Id);
+
+    return status;
+}
+
+Status DeviceEnergyManagementDelegate::StartTimeAdjustRequest(const uint32_t requestedStartTime)
+{
+    Status status = Status::UnsupportedCommand; // Status::Success;
+
+    // TODO: implement
+    DataModel::Nullable<Structs::ForecastStruct::Type> forecast = mDelegate.GetForecast();
+
+    if (!forecast.IsNull())
+    {
+        uint32_t duration = forecast.endTime - forecast.startTime; // the current entire forecast duration
+
+        /* Modify start time and end time */
+        forecast.startTime = requestedStartTime;
+        forecast.endTime   = requestedStartTime + duration;
+
+        mDelegate.SetForecast(forecast);
+    }
```

Comment:
else {
   status = Status::Failure;
}

Reply:
refactored to make this if (forcast.IsNull()) return Failure to make code easier to read.